### ISSUE TITLE
NDEBUG not defined for Release builds

### ISSUE THREAD

**Issue by jfirebaugh** [mapbox/mapbox-gl-native#6887]
The stack trace from #6883 indicates that the `MBGL_CHECK_ERROR` macro was compiled to the error-checking variant in a release build. Most likely, this means that release builds are neglecting to define `NDEBUG`, and therefore including code intended only for debug builds. This includes GL error checking, debug level logging, and `assert`s.

@kkaefer, is this something that may have regressed in the switch to cmake?

**Comment by kkaefer** [comment]
Nope, Release builds are still built with `NDEBUG`. However, it seems like https://github.com/mapbox/mapbox-gl-native/commit/f41546e61456ff85d07c4b21ada54ae7b2709b87 changed the default `BUILDTYPE` for `make ipackage` to `Debug`. Apparently this was in response to https://github.com/mapbox/mapbox-gl-native/pull/5120/commits/9732f02422eb5cd5852d47e888b061aaf6ffe0bd defaulting to `Debug` in the rest of the code as well.

@friedbunny @incanus @1ec5 did you accidentally build iOS packages without `BUILDTYPE=Release`?

**Comment by kkaefer** [comment]
We could add something like this:

```diff
diff --git a/src/mbgl/gl/debugging.hpp b/src/mbgl/gl/debugging.hpp
index 42c31cc..1f1cf09 100644
--- a/src/mbgl/gl/debugging.hpp
+++ b/src/mbgl/gl/debugging.hpp
@@ -2,6 +2,8 @@
 
 #ifndef NDEBUG
 
+#pragma message "WARNING: Building with BUILDTYPE=Debug"
+
 #include <string>
 
 namespace mbgl {
```

**Comment by mikemorris** [comment]
We have a [hard fail](https://github.com/mapbox/mapbox-gl-native/blob/master/platform/node/scripts/after_script.sh#L10) when trying to publish a Node.js package with `BUIILDTYPE=Debug` on Travis or Bitrise - would something like this be helpful?

**Comment by boundsj** [comment]
> @friedbunny @incanus @1ec5 did you accidentally build iOS packages without BUILDTYPE=Release?

@kkaefer we don't use the default when we publish a build. We use a script and it sets the build type to release. See https://github.com/mapbox/mapbox-gl-native/blob/release-ios-v3.4.0/platform/ios/scripts/deploy-packages.sh#L53

**Comment by 1ec5** [comment]
One theory is that we’re building the SDK itself in Release, but it’s linking to an mbgl static library built in Debug.

**Comment by boundsj** [comment]
When I build locally with `SYMBOLS=YES BUILDTYPE=Release make iframework` and `SYMBOLS=YES make iframework` things work as expected in the repro app where there is a zero rect framed map view (as expected means log spew in release and crash in debug). So it does seem like something unexpected happened when I published beta 2. We don't intend to make it possible to publish a debug build using the publish script so I don't think I "ran the script incorrectly" so I will keep digging to figure out what happened.

**Comment by boundsj** [comment]
When I built beta.2 I used Xcode 7.3.1 because https://github.com/mapbox/mapbox-gl-native/issues/6842#issuecomment-257726195. I'm not exactly sure why, but something about `xcode-select`ing that toolchain (I was on Xcode 8 just before that) seems to have caused this Debug output even though the scripts should always result in Release output. 

I spent a bunch of time on this **before deciding to turn my computer off and turn it back on it again**. This appears to have "fixed" the problem. I'm going to close this and will keep an eye on future releases. I think the solution for this already ticketed: https://github.com/mapbox/mapbox-gl-native/issues/2844
